### PR TITLE
Agile crypto

### PR DESCRIPTION
**Pull request by jlmucb**
OK.  It works.

After this checkin, I still have to get the C++ simpleexample working (I havent tried it) and try the other cipher suites but those changes should be small.

The merge should also add roughtime to this repository

### CODE REVIEW

File: go/apps/newfileproxy/fileclient/fileclient.go [hunk 1]
```diff
@@ -84,16 +84,13 @@ func main() {
...
 
 	encryptedFileSecrets, err := ioutil.ReadFile(secretsFileName)
 	if err != nil {
 		rand.Read(fileSecrets)
 	} else {
-		fileSecrets, err = taosupport.Unprotect(clientProgramData.ProgramSymKeys, encryptedFileSecrets)
-		if err != nil {
-			fmt.Printf("fileclient: Error protecting data\n")
-		}
+		fileSecrets, err = clientProgramData.ProgramCryptingKey.Decrypt(encryptedFileSecrets)
```
Comment (sidtelang) [inline go/apps/newfileproxy/fileclient/fileclient.go]: check for non-nil err?

File: go/apps/simpleexample/SimpleServer/simpleserver.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ import (
 
 	"github.com/jlmucb/cloudproxy/go/apps/simpleexample/common"
 	taosupport "github.com/jlmucb/cloudproxy/go/support_libraries/tao_support"
-	"github.com/jlmucb/cloudproxy/go/tao"
+	// "github.com/jlmucb/cloudproxy/go/tao"
```
Comment (sidtelang) [inline go/apps/simpleexample/SimpleServer/simpleserver.go]: remove?

File: go/tao/guard.go [hunk 3]
```diff
@@ -20,7 +20,7 @@ import (
 	"errors"
 
 	"github.com/jlmucb/cloudproxy/go/tao/auth"
-	"github.com/jlmucb/cloudproxy/go/util"
+	// "github.com/jlmucb/cloudproxy/go/util"
```
Comment (sidtelang) [inline go/tao/guard.go]: remove?

File: go/tao/keys.go [hunk 4]
```diff
@@ -15,1544 +15,1637 @@
...
-func MarshalDeriverProto(d *Deriver) (*CryptoKey, error) {
-	k := marshalHMACSHADerivingKeyV1(d)
+// PrepareX509Template fills out an X.509 template for use in x509.CreateCertificate.
+func PrepareX509Template(pkAlg int, sigAlg int, sn int64, subjectName *pkix.Name) *x509.Certificate {
+	return &x509.Certificate{
+		SignatureAlgorithm: x509.ECDSAWithSHA256,
+		PublicKeyAlgorithm: x509.ECDSA,
+		Version:            2, // x509v3
+		// It's always allowed for self-signed certs to have serial 1.
+		SerialNumber: new(big.Int).SetInt64(1),
```
Comment (sidtelang) [inline go/tao/keys.go]: this ignores the argument and always uses serial number 1

File: go/apps/host/host.go [hunk 5]
```diff
@@ -35,6 +35,7 @@ import (
 	"github.com/jlmucb/cloudproxy/go/tao"
 	"github.com/jlmucb/cloudproxy/go/util"
 	"github.com/jlmucb/cloudproxy/go/util/options"
+	// "github.com/golang/crypto/ssh/terminal"
```
Comment (jlmucb) [inline go/apps/host/host.go]: yes, I have to switch to this to make it work on the mac

File: go/apps/newfileproxy/fileclient/fileclient.go (hunk 1 above)
Comment (jlmucb) [inline go/apps/newfileproxy/fileclient/fileclient.go]: thanks.  change made

**Pull request merged**